src/python/pants/backend/kotlin/compile/kotlinc_test.py

```diff
@@ -0,0 +1,583 @@
+    )
+
+
+@maybe_skip_jdk_test
+def test_compile_jdk_12_file_fails_on_jdk_11(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "BUILD": dedent(
+                """\
+                kotlin_sources(
+                    name = 'lib',
+                    jdk = 'adopt:1.11',
+                )
+                """
+            ),
+            "3rdparty/jvm/BUILD": DEFAULT_KOTLIN_STDLIB_TARGET,
+            "3rdparty/jvm/default.lock": DEFAULT_LOCKFILE,
+            "ExampleLib.kt": KOTLIN_LIB_JDK12_SOURCE,
+        }
+    )
+    coarsened_target = expect_single_expanded_coarsened_target(
+        rule_runner, Address(spec_path="", target_name="lib")
+    )
+
+    with pytest.raises(ExecutionError):
+        rule_runner.request(
+            RenderedClasspath,
+            [
+                CompileKotlinSourceRequest(
+                    component=coarsened_target, resolve=make_resolve(rule_runner)
+                )
+            ],
+        )
+
+
+@maybe_skip_jdk_test
+def test_compile_with_deps(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "BUILD": dedent(
+                """\
+                kotlin_sources(
+                    name = 'main',
+                    dependencies = [
+                        'lib:lib',
+                        # TODO: Remove this once kotlin stdlib dep is injected.
+                        "3rdparty/jvm:org.jetbrains.kotlin_kotlin-stdlib_1.6.0",
+                    ]
+                )
+                """
+            ),
+            "3rdparty/jvm/BUILD": DEFAULT_KOTLIN_STDLIB_TARGET,
+            "3rdparty/jvm/default.lock": DEFAULT_LOCKFILE,
+            "Example.kt": KOTLIN_LIB_MAIN_SOURCE,
+            "lib/BUILD": dedent(
+                """\
+                kotlin_sources(
+                    name = 'lib',
+                    dependencies = [
+                        # TODO: Remove this once kotlin stdlib dep is injected.
+                        "3rdparty/jvm:org.jetbrains.kotlin_kotlin-stdlib_1.6.0",
+                    ],
+                )
+                """
+            ),
+            "lib/ExampleLib.kt": KOTLIN_LIB_SOURCE,
+        }
+    )
+    classpath = rule_runner.request(
+        RenderedClasspath,
+        [
+            CompileKotlinSourceRequest(
+                component=expect_single_expanded_coarsened_target(
+                    rule_runner, Address(spec_path="", target_name="main")
+                ),
+                resolve=make_resolve(rule_runner),
+            )
+        ],
+    )
+    assert classpath.content == {
+        ".Example.kt.main.kotlin.jar": {
+            "META-INF/MANIFEST.MF",
+            "META-INF/main.kotlin_module",
+            "org/pantsbuild/example/ExampleKt.class",
+        }
+    }
+
+
+@maybe_skip_jdk_test
+@pytest.mark.xfail(reason="compilation succeeded for some reason")
```

Comment:
This compilation should have failed, but instead succeeds. Unclear to me why.

@chrisjrn: Thoughts?

Reply:
Would rebase onto `main` post #14934, and add the `@logging` decorator to this method to get more info. But maybe look at which classpath entries it is actually getting, to see whether it is receiving more than it should.

Reply:
Copy-pasta error: `Example.scala` -> `Example.kt` fixes it.